aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1359,127 +1356,110 @@ Tensor orgqr(const Tensor& input, const Tensor& tau) {
-  auto batch_size = batchCount(v);
-  auto n = v.size(-1);
-  auto lda = std::max(int64_t{1}, n);
+/*
+  Computes eigenvalues and eigenvectors of the tensor 'input'.
 
-  // NumPy allows lowercase input for UPLO argument
-  // It is assumed that uplo_str is either "U" or "L"
-  char uplo = std::toupper(uplo_str[0]);
-  char jobz = compute_v ? 'V' : 'N';
+  Args:
+  * 'input' - input Tensor for eigendecomposition
+  * 'values' - Tensor to store computed eigenvalues
+  * 'vectors' - Tensor to store computed eigenvectors
+  * 'infos' - Tensor to store LAPACK/MAGMA/cuSOLVER error codes
+  * 'compute_eigenvectors' - controls whether eigenvectors should be computed
+  * 'uplo_str' - controls the portion of input matrix to consider in computations, allowed values are "u", "U", "l", "L"
+*/
+std::tuple<Tensor&, Tensor&> linalg_eigh_out_info(
+    const Tensor& input,
+    Tensor& values,
+    Tensor& vectors,
+    Tensor& infos,
+    bool compute_eigenvectors,
+    const std::string& uplo_str) {
+  // These internal asserts make explicit the assumptions in the implementation
+  // Error check with the actual error messages are done on the higher level of
+  // the hierarchy of calls
+  TORCH_INTERNAL_ASSERT(input.dim() >= 2);
+  TORCH_INTERNAL_ASSERT(input.size(-2) == input.size(-1));
 
-  // Using 'int' instead of int32_t or int64_t is consistent with the current LAPACK interface
-  // It really should be changed in the future to something like lapack_int that depends on the specific LAPACK library that is linked
-  // or switch to supporting only 64-bit indexing by default.
-  int info;
-  int lwork = -1;
-  int lrwork = -1;
-  int liwork = -1;
-  scalar_t work_query;
-  value_t rwork_query;
-  int iwork_query;
+  TORCH_INTERNAL_ASSERT(input.device() == vectors.device());
+  TORCH_INTERNAL_ASSERT(input.device() == values.device());
 
-  // Run lapackSyevd once, first to get the optimum work size.
-  // Since we deal with batches of matrices with the same dimensions, doing this outside
-  // the main loop saves (batch_size - 1) workspace queries which would provide the same result
-  // and (batch_size - 1) calls to allocate and deallocate workspace using at::empty()
-  lapackSyevd<scalar_t, value_t>(jobz, uplo, n, v_data, lda, w_data, &work_query, lwork, &rwork_query, lrwork, &iwork_query, liwork, &info);
+  // eigenvalues are always real-valued
+  ScalarType real_dtype = toValueType(input.scalar_type());
+  TORCH_INTERNAL_ASSERT(values.scalar_type() == real_dtype);
+  TORCH_INTERNAL_ASSERT(input.scalar_type() == vectors.scalar_type());
 
-  lwork = std::max<int>(1, real_impl<scalar_t, value_t>(work_query));
-  Tensor work = at::empty({lwork}, v.options());
-  liwork = std::max<int>(1, iwork_query);
-  Tensor iwork = at::empty({liwork}, at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.device() == input.device());
+  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
```

Comment:
This is interesting that it's checking the number of elements and not the shape of infos. Should it just take the shape?

Reply:
`infos` can have the shape equal to `input.shape[:-2]` or `(batchCount(input), )`, both would work with the current implementation, the important requirement is being contiguous.
`input.shape[:-2]` shape might be useful in the future for easier checking the error code for the specific matrix in batched input when we would have a user-exposed way to get `infos` tensor. 
1-dimensional tensor of shape `(batchCount(input), )` is currently used for the internal implementation everywhere.

Reply:
OK. Would you add a comment explaining that thinking?